staging/src/k8s.io/client-go/transport/cert_rotation.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package transport
+
+import (
+	"bytes"
+	"crypto/tls"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/connrotation"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+const workItemKey = "key"
+
+// CertCallbackRefreshDuration is exposed so that integration tests can crank up the reload speed.
+var CertCallbackRefreshDuration = 5 * time.Minute
+
+type dynamicClientCert struct {
+	clientCert *tls.Certificate
+	certMtx    sync.RWMutex
+
+	reload     func(*tls.CertificateRequestInfo) (*tls.Certificate, error)
+	connDialer *connrotation.Dialer
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+}
+
+func certRotatingDialer(tlsConfig *tls.Config, dial utilnet.DialFunc, stopCh <-chan struct{}) utilnet.DialFunc {
+	d := &dynamicClientCert{
+		reload:     tlsConfig.GetClientCertificate,
+		connDialer: connrotation.NewDialer(connrotation.DialFunc(dial)),
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "DynamicClientCertificate"),
```

Comment:
I thought colliding names produced errors.

Reply:
I don't think so, looking at the code the metrics will get shared if there are multiple queues with the same name
I could be wrong!
We could try to add the cert file names here, but technically we don't know them and there may be a GetCert function in which case we don't really have any kind of identifier beyond a pointer

Reply:
Name is only used for metrics AFAIK. Multiple queue's with same name just get aggreted in queue metrics which isn't terrible. Nit: traditionally the names have been lower snake case, but I see we've diverged on that convention for some of the other cert stuff.